proto/buf/alpha/registry/v1alpha1/webhook_payload.proto

```diff
@@ -0,0 +1,52 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package buf.alpha.registry.v1alpha1;
+
+import "buf/alpha/registry/v1alpha1/webhook.proto";
+import "google/protobuf/timestamp.proto";
+
+// WebhookListenerService represents a service implemented by a customer, to
+// receive webhook events invocations from the BSR.
+service WebhookListenerService {
+  // ListenWebhook listens events from a registered webhook, triggered in the
+  // BSR and forwared to the customer's backend.
+  rpc ListenWebhook(ListenWebhookRequest) returns (ListenWebhookResponse) {}
+}
+
+// ListenWebhookRequest is the request sent to the customer's webhook listener
+// service on each webhook invocation.
+message ListenWebhookRequest {
+  // payload is the actual webhook payload.
+  WebhookPayload payload = 1;
+}
+
+// ListenWebhookResponse empty response for successful invocations receipts.
+message ListenWebhookResponse {}
+
+// WebhookPayload defines the structure of an actual webhook invocation.
+message WebhookPayload {
+  // The timestamp of when the event happened.
+  google.protobuf.Timestamp event_time = 1;
+  // The event that happened.
+  buf.alpha.registry.v1alpha1.WebhookEvent event = 2;
+  // The repository id tied to the event.
+  string repository_id = 3;
+  // Te repository commit id tied to the event.
+  string repository_commit_id = 4;
+  // The user id of the author who triggered the event.
+  string author_user_id = 5;
```

Comment:
I'm thinking maybe the names would be more helpful here than the ids since the user wouldn't have a way to look up these ids themselves

Reply:
In the context of an UI-oriented use case, sure, human-readable attributes are much more helpful (so they're rendered and seen by a human). But webhooks will be consumed by their backend service, and most likely they already have these IDs stored. Repo names and owner names can change, so that also leave us with possible race conditions. In general terms, it's much easier to deal with IDs for relational objects, those will never change.

Reply:
Another option is to inject the complete commit object, something like:

```proto
message WebhookPayload {
  google.protobuf.Timestamp event_time = 1;
  buf.alpha.registry.v1alpha1.WebhookEvent event = 2;
  RepositoryCommit repository_commit = 3; // which would include already the repository and author ids
  Repository repository = 4; // an overkill in my opinion, the event relates to the commit, not to the repository
}
```

Reply:
Including more information seems like the right direction. Looking at something like https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads, the response payloads seem quite detailed. The two things that I can come up with why that seems desirable:

- If we emit webhooks for an event like "repo rename", you'd probably want to include details like old/new name.
- If in response to receiving a webhook users would need to send `n` more requests to have a picture of what happened, besides the ux being bad there is some potential for load amplification when pushing out webhooks (not sure if that will really be a concern, the ux argument is probably stronger).

How detailed are the payloads for other popular webhooks (slack, github, others)?

Reply:
True! and that makes me think, we should have a different RPC (with different payloads), per event type, something like `ListenRepositoryPushEvent`. So we can customize how/which data we want to send to the customer, regardless of the event.

Downside of this proposal, is that if the customer wants to subscribe to just 1 event, but we support 20, they'll endup with 20 autogenerated functions that do nothing. Doesn't seem too bad imo, depending on their language they can just delete them.

Reply:
Having a toplevel `oneof` in the webhook payload to separate out diffent events as compared to having different service names feels closer to what I'm reading in the github docs on webhooks. Related to the pkg in the url discussion, I don't think it'd be great if we required users to expose a new handler/path for each event type they'd subscribe on.

We had something similar to that in the now defunct audit log structure:

https://github.com/bufbuild/buf/blob/14ed91784fcd5cd06763c420ce6d765092571430/proto/buf/alpha/audit/v1alpha1/envelope.proto#L579-L583

And I've seen it before in bazel as well: https://github.com/bazelbuild/bazel/blob/2594ea48f65008e81d29bbf77b5798f4c43fc454/src/main/java/com/google/devtools/build/lib/buildeventstream/proto/build_event_stream.proto#L1119-L1122

Reply:
Done